### PR TITLE
Location settings toggle

### PR DESCRIPTION
**Pull request by catt-stefano**
Toggle in setting shows opt-in status.
Enabling toggle leads to onboarding.

### CODE REVIEW

File: app/src/main/java/net/squanchy/onboarding/location/LocationOnboardingActivity.java [hunk 1]
```diff
@@ -48,21 +50,25 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+        optInPersister = component.optInPersister();
 
         googleApiClient.connect();
 
         setContentView(R.layout.activity_location_onboarding);
         contentRoot = findViewById(R.id.onboarding_content_root);
 
-        findViewById(R.id.skip_button).setOnClickListener(view -> markPageAsSeenAndFinish());
+        findViewById(R.id.skip_button).setOnClickListener(view -> {
+            optInPersister.setOptInPreferenceTo(false);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/onboarding/location/LocationOnboardingActivity.java]: Extract a method here please for these two calls? something like `optOutFromProximity` so it mirror the other case maybe even

File: app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package net.squanchy.proximity.preconditions;
+
+import android.content.SharedPreferences;
+
+public class OptInPreferencePersister {
+
+    private static final String OPT_IN_PREFERENCE_KEY = "opt_in_preference";
+    private final SharedPreferences preferences;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java]: Empty line between these two please

File: app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package net.squanchy.proximity.preconditions;
+
+import android.content.SharedPreferences;
+
+public class OptInPreferencePersister {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java]: Can we call this `ProximityOptInPersister` for consistency with the other similar classes?

File: app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static final String OPT_IN_PREFERENCE_KEY = "opt_in_preference";
+    private final SharedPreferences preferences;
+
+    public OptInPreferencePersister(SharedPreferences preferences) {
+        this.preferences = preferences;
+    }
+
+    public void setOptInPreferenceTo(boolean granted) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java]: I would rather have two public methods for the two cases, with more descriptive names, such as: `storeUserOptedIn()` e `storeUserOptedOut()`?

Using a boolean flag as the only parameter in APIs is a discouraged practice. The [alternative](https://books.google.co.uk/books?id=ka2VUBqHiWkC&lpg=PA190&ots=yZLgJjr_U-&pg=PA190#v=onepage&q&f=false) would be using an enum with two values (e.g., `storeUserPreference(OptIn optIn)` with two possible values, `OPTED_IN` and `OPTED_OUT`). 

I'd rather have the two methods though.

File: app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        preferences.edit().putBoolean(OPT_IN_PREFERENCE_KEY, granted).apply();
+    }
+
+    public boolean getOptInPreferenceGranted() {
+        return preferences.getBoolean(OPT_IN_PREFERENCE_KEY, true);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java]: Can you extract the `true` as a constant that makes it clear it's a default-enabled thing? E.g., `DEFAULT_OPTED_IN`

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java [hunk 6]
```diff
@@ -20,12 +24,17 @@
 
 public class SettingsFragment extends PreferenceFragment {
 
+    public static final int TURN_LOCATION_ON_REQUEST_CODE = 6429;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/settings/SettingsFragment.java]: Usual name format is: `REQUEST_...`

File: app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    public OptInPreferencePersister(SharedPreferences preferences) {
+        this.preferences = preferences;
+    }
+
+    public void setOptInPreferenceTo(boolean granted) {
+        preferences.edit().putBoolean(OPT_IN_PREFERENCE_KEY, granted).apply();
+    }
+
+    public boolean getOptInPreferenceGranted() {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/proximity/preconditions/OptInPreferencePersister.java]: We don't usually call getters `getX`. Can you rename to `userOptedIn()`?

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java [hunk 8]
```diff
@@ -75,6 +90,8 @@ private void removeDebugCategory() {
     public void onStart() {
         super.onStart();
 
+        locationPreferences.setChecked(optInPreferencePersister.getOptInPreferenceGranted());
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/settings/SettingsFragment.java]: Here would also be the place where you check the remote config and disable the preference altogether if the kill switch is active. Next PR?

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java [hunk 9]
```diff
@@ -124,6 +141,17 @@ private void onSignedOut() {
         );
     }
 
+    private boolean handleLocationPreferenceChange(Boolean isEnabling) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/settings/SettingsFragment.java]: Parameter could be `boolean`

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java [hunk 10]
```diff
@@ -124,6 +141,17 @@ private void onSignedOut() {
...
+        if (isEnabling) {
+            navigator.toOnboardingForResult(OnboardingPage.LOCATION, TURN_LOCATION_ON_REQUEST_CODE);
+            return false;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/settings/SettingsFragment.java]: ⚠️ this should return true regardless, it's the "change handled" value.

File: app/src/main/res/xml/settings_preferences.xml [hunk 11]
```diff
@@ -34,7 +34,7 @@
     <SwitchPreference
       android:key="@string/location_preference_key"
       android:defaultValue="false"
-      android:enabled="false"
+      android:enabled="true"
```
Comment (rock3r) [inline app/src/main/res/xml/settings_preferences.xml]: I don't think this should matter if things are working properly.

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -109,6 +109,7 @@
   <string name="onboarding_error_location_failed">Unable to verify location settings for the device.</string>
   <string name="onboarding_error_location_failed_action">ENABLE MANUALLY</string>
   <string name="onboarding_error_google_client_connection">Unable to verify location settings for the device. Please try again shortly.</string>
+  <string name="onboarding_error_canceled_contest">The contest is unavailable.</string>
```
Comment (rock3r) [inline app/src/main/res/values/strings.xml]: This seems unused.

**Review by rock3r — CHANGES_REQUESTED** [review]
Good one but there's some comments to address

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java (hunk 10 above)
Comment (catt-stefano) [reply app/src/main/java/net/squanchy/settings/SettingsFragment.java]: If we return true, the toggle animation begins while opening the onboarding screen, even if we are going to skip or dismiss the onboarding screen (then the toggle animates back off after the onStart()).
Are we okay with this side effect?

**Review by rock3r — APPROVED** [review]
Nice! 👍

File: app/src/main/java/net/squanchy/settings/SettingsFragment.java (hunk 10 above)
Comment (rock3r) [reply app/src/main/java/net/squanchy/settings/SettingsFragment.java]: Will sort it out when I add the "show standings" preference

**Pull request merged**